flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/MultiStateKeyIterator.java

```diff
@@ -31,13 +31,22 @@
 import java.io.IOException;
 import java.util.Iterator;
 import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.Stream;
 
 /**
  * An iterator for reading all keys in a state backend across multiple partitioned states.
  *
  * <p>To read unique keys across all partitioned states callers must invoke {@link
  * MultiStateKeyIterator#remove}.
  *
+ * <p>Note: This is a replacement of the original implementation which used streams with a known
+ * flaw in the {@link Stream#flatMap(java.util.function.Function)} implementation that lead to
+ * completely enumerating and buffering nested iterators event for a single call to {@link
+ * MultiStateKeyIterator#hasNext}.
+ *
+ * @see <a
+ *     href="https://bugs.openjdk.org/browse/JDK-8267359">https://bugs.openjdk.org/browse/JDK-8267359</a>
```

Comment:
IMO, this comment about why we update the logic is not necessary, which could be found in the Jira Ticket.
Or you could just add simple description before outerIter, e.g. : "Avoid using Stream#flatMap due to xxx, see FLINK-26585 for more details"